stablehlo/conversions/tosa/transforms/StablehloQuantLegalizeToTosaRescale.cpp

```diff
@@ -0,0 +1,257 @@
+
+#include "mlir/Dialect/Func/IR/FuncOps.h"
+#include "mlir/Dialect/Quant/QuantOps.h"
+#include "mlir/Dialect/Tosa/IR/TosaOps.h"
+#include "mlir/Dialect/Tosa/Utils/ConversionUtils.h"
+#include "mlir/Dialect/Tosa/Utils/QuantUtils.h"
+#include "mlir/IR/BuiltinAttributes.h"
+#include "mlir/IR/PatternMatch.h"
+#include "mlir/Parser/Parser.h"
+#include "mlir/Support/LogicalResult.h"
+#include "mlir/Transforms/GreedyPatternRewriteDriver.h"
+#include "stablehlo/conversions/tosa/transforms/Passes.h"
+#include "stablehlo/dialect/StablehloOps.h"
+
+#define PASS_NAME "stablehlo-quant-legalize-to-tosa-rescale"
+#define DEBUG_TYPE PASS_NAME
+
+using namespace mlir;
+using namespace mlir::tosa;
+
+namespace mlir {
+namespace tosa {
+
+#define GEN_PASS_DEF_STABLEHLOQUANTLEGALIZETOTOSARESCALEPASS
+#include "stablehlo/conversions/tosa/transforms/Passes.h.inc"
+
+namespace {
+
+struct StablehloQuantLegalizeToTosaRescalePass
+    : impl::StablehloQuantLegalizeToTosaRescalePassBase<
+          StablehloQuantLegalizeToTosaRescalePass> {
+  void runOnOperation() final;
+
+  LogicalResult initialize(MLIRContext* ctx) override;
+
+ private:
+  FrozenRewritePatternSet patterns;
+};
+
+#define DECL_CONVERT_OP(stablehlo_op)                                        \
+  struct ConvertStablehloQuant##stablehlo_op##Op                             \
+      : public OpRewritePattern<stablehlo::stablehlo_op##Op> {               \
+    using OpRewritePattern<stablehlo::stablehlo_op##Op>::OpRewritePattern;   \
+    LogicalResult matchAndRewrite(stablehlo::stablehlo_op##Op op,            \
+                                  PatternRewriter& rewriter) const override; \
+  }
+
+DECL_CONVERT_OP(Add);
+DECL_CONVERT_OP(Subtract);
+
+#undef DECL_CONVERT_OP
+
+// create a tosa rescale op and return its result value
+Value buildRescale(PatternRewriter& rewriter, Location loc,
+                   ShapedType output_type, Value input_val, int32_t multiplier,
+                   int32_t shift, int64_t inputZp, int64_t outputZp,
+                   bool double_round, bool scale32, bool perChannel) {
+  // bool input_unsigned = input_val.getType().isUnsignedInteger();
+  // bool output_unsigned = output_type.isUnsignedInteger();
+
+  auto rescale_op = rewriter.create<tosa::RescaleOp>(
+      loc, output_type, input_val,
+      rewriter.getI32IntegerAttr(static_cast<int32_t>(inputZp)),
+      rewriter.getI32IntegerAttr(static_cast<int32_t>(outputZp)),
+      rewriter.getDenseI32ArrayAttr({multiplier}),
+      rewriter.getDenseI8ArrayAttr({static_cast<int8_t>(shift)}),
+      rewriter.getBoolAttr(scale32), rewriter.getBoolAttr(double_round),
+      rewriter.getBoolAttr(perChannel));
+
+  return rescale_op.getResult();
+}
+
+// Creates TOSA rescale op with int32 output
+Value buildRescaleToInt32(PatternRewriter& rewriter, Location loc,
+                          Value inputVal, double inputScale, int64_t inputZp) {
+  auto inputType = dyn_cast<mlir::ShapedType>(inputVal.getType());
```

Comment:
replace `dyn_cast` with `cast` and remove the following check.

Reply:
done